### PR TITLE
Update React Router to version 4.0

### PR DESCRIPTION
**Pull request by tkrotoff**
Update React Router definitions to version 4.0
See https://reacttraining.com/react-router/web/guides/quick-start

Tested with:
```
./node_modules/.bin/tsc -p react-router
./node_modules/.bin/tsc -p react-router-dom
npm run lint react-router
npm run lint react-router-dom
```

I've added all examples from React Router website, converted them to TypeScript in order to test `index.d.ts`. ~~Problem: [types-publisher](https://github.com/Microsoft/types-publisher/issues/318) wants to be too smart~~

- [X] Make your PR against the `master` branch.
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `tsc` without errors.
- [X] Run `npm run lint package-name` if a `tslint.json` is present.

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://reacttraining.com/react-router/web/guides/quick-start
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

### CODE REVIEW

**Comment by tkrotoff** [comment]
Failing Travis CI with `Error: In /home/travis/build/DefinitelyTyped/DefinitelyTyped/react-router: Expected file 'examples/Ambiguous.tsx' to be named react-router-tests.ts` is due to https://github.com/Microsoft/types-publisher/issues/318

File: react-router-dom/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// Type definitions for React Router 4.0
+// Project: https://github.com/ReactTraining/react-router
+// Definitions by: Tanguy Krotoff <https://github.com/tkrotoff>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
```
Comment (yuit) [inline react-router-dom/index.d.ts]: nit: could you add typescript version ? "// TypeScript Version: <version>

File: react-router/index.d.ts [hunk 2]
```diff
@@ -1,62 +1,104 @@
...
-//                 Václav Ostrožlík <https://github.com/vasek17>
-//                 Nathan Brown <https://github.com/ngbrown>
-//                 Alex Wendland <https://github.com/awendland>
-//                 Kostya Esmukov <https://github.com/KostyaEsmukov>
-//                 John Reilly <https://github.com/johnnyreilly>
-//                 Karol Janyst <https://github.com/LKay>
-//                 Dovydas Navickas <https://github.com/DovydasNavickas>
+// Type definitions for React Router 4.0
+// Project: https://github.com/ReactTraining/react-router
+// Definitions by: Sergey Buturlakin <https://github.com/sergey-buturlakin>, Yuichi Murata <https://github.com/mrk21>, Václav Ostrožlík <https://github.com/vasek17>, Nathan Brown <https://github.com/ngbrown>, Alex Wendland <https://github.com/awendland>, Kostya Esmukov <https://github.com/KostyaEsmukov>, Tanguy Krotoff <https://github.com/tkrotoff>
```
Comment (yuit) [inline react-router/index.d.ts]: Same add "// TypeScript Version: <version>"

Comment (yuit) [inline react-router/index.d.ts]: Question: why is this file change

**Review by yuit — COMMENTED** [review]
is the v3 for React Router 3.0?

**Comment by yuit** [comment]
I put the reviewer tag on as there is dependent on the type publisher issue

File: react-router-dom/index.d.ts (hunk 1 above)
Comment (tkrotoff) [inline react-router-dom/index.d.ts]: @yuit See commit https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15170/commits/ee0f49ec1cd4b63aab4bbb630cb389db68f7e406

**Comment by tkrotoff** [comment]
@yuit you can remove the reviewer tag, the types-publisher issue is solved (only a documentation problem)

**Comment by donaldpipowitch** [comment]
@tkrotoff are `url`, `isExact` and `path` in [`match`](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/c4dddc32c16848cb8065c7f9df95d88fba3d721e#diff-0f71fe6a8f24d2e8195a1b467fe555cdR85) really optional (~can be `undefined`)?

* `path` looks to [default to `'/'`](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/matchPath.js#L33).
* `isExact` must be [`true` or `false`](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/matchPath.js#L41).
* `url` was [either matched](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/matchPath.js#L40) or is [`'/'`](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/matchPath.js#L48).

**Comment by tkrotoff** [comment]
@donaldpipowitch probably room for improvements, please send a PR since this one has been already merged

**Comment by donaldpipowitch** [comment]
Just wanted to be sure that this is actually wrong, before I make a PR :) thanks

**Comment by paulvanbrenk** [comment]
@donaldpipowitch ping me when you have a PR ready, and I'll merge it in. Thanks!

**Comment by donaldpipowitch** [comment]
I made a PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15314 :)

**Pull request merged**